mesonbuild/optinterpreter.py

```diff
@@ -217,34 +151,82 @@ def evaluate_statement(self, node: mparser.BaseNode) -> None:
         if func_name != 'option':
             raise OptionException('Only calls to option() are allowed in option files.')
         (posargs, kwargs) = self.reduce_arguments(node.args)
-
-        if len(posargs) != 1:
-            raise OptionException('Option() must have one (and only one) positional argument')
-        opt_name = posargs[0]
-        if not isinstance(opt_name, str):
-            raise OptionException('Positional argument must be a string.')
+        self.func_option(posargs, kwargs)
+
+    @typed_kwargs('option',
+                  KwargInfo('type', str, required=True),
+                  KwargInfo('description', str, default=''),
+                  KwargInfo('yield', bool, default=coredata.default_yielding, since='0.45.0'),
+                  KwargInfo('choices', (ContainerTypeInfo(list, str), type(None))),
+                  KwargInfo('value', object),
+                  KwargInfo('min', (int, type(None))),
+                  KwargInfo('max', (int, type(None))),
+                  KwargInfo('deprecated', (bool, ContainerTypeInfo(dict, str), ContainerTypeInfo(list, str)),
+                            default=False, since='0.60.0')
+                  )
+    @typed_pos_args('option', str)
+    def func_option(self, args: T.Tuple[str], kwargs: 'FuncOptionArgs') -> None:
+        opt_name = args[0]
         if optname_regex.search(opt_name) is not None:
             raise OptionException('Option names can only contain letters, numbers or dashes.')
         if is_invalid_name(opt_name):
             raise OptionException('Option name %s is reserved.' % opt_name)
-        key = mesonlib.OptionKey(opt_name, self.subproject)
-
-        if 'yield' in kwargs:
-            FeatureNew.single_use('option yield', '0.45.0', self.subproject)
 
-        if 'type' not in kwargs:
-            raise OptionException('Option call missing mandatory "type" keyword argument')
-        opt_type = kwargs.pop('type')
-        if not isinstance(opt_type, str):
-            raise OptionException('option() type must be a string')
-        if opt_type not in option_types:
-            raise OptionException('Unknown type %s.' % opt_type)
+        opt_type = kwargs['type']
+        parser = self.option_types.get(opt_type)
+        if not parser:
+            raise OptionException(f'Unknown type {opt_type}.')
+        description = kwargs['description'] or opt_name
 
-        description = kwargs.pop('description', '')
-        if not isinstance(description, str):
-            raise OptionException('Option descriptions must be strings.')
+        # Only keep in kwargs arguments that are used by option type's parser
+        # because they use @permittedKwargs().
+        known_parser_kwargs = {'value', 'choices', 'yield', 'min', 'max'}
+        parser_kwargs = {k: v for k, v in kwargs.items() if k in known_parser_kwargs and v is not None}
```

Comment:
you can resolve the last mypy error either by forcing the type (I think) or casting (for sure). I think:
```python
parser_kwargs: 'TypeDictUsedByParser' = {k: v for ...}
```
will work, but you might hav to do:
```python
opt = parser(description, T.cast('TypeDictUsedByParser', parser_kwargs))
```

Reply:
> parser_kwargs: 'TypeDictUsedByParser' = {k: v for ...}

That did not work, but the cast worked. Thanks.